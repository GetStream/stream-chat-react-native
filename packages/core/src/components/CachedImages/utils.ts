export const extractPathname = (uri: string | undefined) => {
  const parsedUrl = uri?.split('?')?.[0];
  // Extracts pathname. Need to use a regex cause url.parse is deprecated and RN doesnt have
  // URL constructor. Replace / with ___ to avoid creating folders
  return parsedUrl?.replace(/(https?:\/\/)?/, '').replace(/\//g, '___');
};
