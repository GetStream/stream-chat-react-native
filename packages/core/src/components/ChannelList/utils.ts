import type { Channel, StreamChat } from 'stream-chat';

import type {
  DefaultAttachmentType,
  DefaultChannelType,
  DefaultCommandType,
  DefaultEventType,
  DefaultMessageType,
  DefaultReactionType,
  DefaultUserType,
  UnknownType,
} from '../../types/types';

type MoveParameters<
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType,
> = {
  channels: Channel<At, Ch, Co, Ev, Me, Re, Us>[];
  cid: string;
};

export const moveChannelUp = <
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType,
>({
  cid,
  channels = [],
}: MoveParameters<At, Ch, Co, Ev, Me, Re, Us>) => {
  // get channel from channels
  const index = channels.findIndex((c) => c.cid === cid);
  if (index <= 0) return channels;
  const channel = channels[index];

  // remove channel from current position and add to start
  channels.splice(index, 1);
  channels.unshift(channel);

  return [...channels];
};

type GetParameters<
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType,
> = {
  client: StreamChat<At, Ch, Co, Ev, Me, Re, Us>;
  id: string;
  type: string;
};

export const getChannel = async <
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType,
>({
  client,
  id,
  type,
}: GetParameters<At, Ch, Co, Ev, Me, Re, Us>) => {
  const channel = client.channel(type, id);
  await channel.watch();
  return channel;
};

export const DEFAULT_QUERY_CHANNELS_LIMIT = 10;
export const MAX_QUERY_CHANNELS_LIMIT = 30;
