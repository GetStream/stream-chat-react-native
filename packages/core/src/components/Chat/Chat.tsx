import React, { PropsWithChildren, useEffect, useState } from 'react';
import { Platform } from 'react-native';
import Dayjs from 'dayjs';

import { useCreateChatContext } from './hooks/useCreateChatContext';
import { useIsOnline } from './hooks/useIsOnline';
import { useMutedUsers } from './hooks/useMutedUsers';

import { ChatContextValue, ChatProvider } from '../../contexts/chatContext/ChatContext';
import { useOverlayContext } from '../../contexts/overlayContext/OverlayContext';
import { DeepPartial, ThemeProvider } from '../../contexts/themeContext/ThemeContext';
import {
  TranslationContextValue,
  TranslationProvider,
} from '../../contexts/translationContext/TranslationContext';
import { useStreami18n } from '../../hooks/useStreami18n';

import { SDK } from '../../native';
import { version } from '../../version.json';

import type { Channel } from 'stream-chat';

import type { Theme } from '../../contexts/themeContext/utils/theme';
import type { Streami18n } from '../../utils/Streami18n';
import type {
  DefaultAttachmentType,
  DefaultChannelType,
  DefaultCommandType,
  DefaultEventType,
  DefaultMessageType,
  DefaultReactionType,
  DefaultUserType,
  UnknownType,
} from '../../types/types';

export type ChatProps<
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType,
> = Pick<ChatContextValue<At, Ch, Co, Ev, Me, Re, Us>, 'client'> & {
  /**
   * When false, ws connection won't be disconnection upon backgrounding the app.
   * To receive push notifications, its necessary that user doesn't have active
   * websocket connection. So by default, we disconnect websocket connection when
   * app goes to background, and reconnect when app comes to foreground.
   */
  closeConnectionOnBackground?: boolean;
  /**
   * Instance of Streami18n class should be provided to Chat component to enable internationalization.
   *
   * Stream provides following list of in-built translations:
   * 1. English (en)
   * 2. Dutch (nl)
   * 3. ...
   * 4. ...
   *
   * Simplest way to start using chat components in one of the in-built languages would be following:
   *
   * ```
   * const i18n = new Streami18n('nl');
   * <Chat client={chatClient} i18nInstance={i18n}>
   *  ...
   * </Chat>
   * ```
   *
   * If you would like to override certain keys in in-built translation.
   * UI will be automatically updated in this case.
   *
   * ```
   * const i18n = new Streami18n('nl');
   *
   * i18n.registerTranslation('nl', {
   *  'Nothing yet...': 'Nog Niet ...',
   *  '{{ firstUser }} and {{ secondUser }} are typing...': '{{ firstUser }} en {{ secondUser }} zijn aan het typen...',
   * });
   *
   * <Chat client={chatClient} i18nInstance={i18n}>
   *  ...
   * </Chat>
   * ```
   *
   * You can use the same function to add whole new language.
   *
   * ```
   * const i18n = new Streami18n('it');
   *
   * i18n.registerTranslation('it', {
   *  'Nothing yet...': 'Non ancora ...',
   *  '{{ firstUser }} and {{ secondUser }} are typing...': '{{ firstUser }} a {{ secondUser }} stanno scrivendo...',
   * });
   *
   * // Make sure to call setLanguage to reflect new language in UI.
   * i18n.setLanguage('it');
   * <Chat client={chatClient} i18nInstance={i18n}>
   *  ...
   * </Chat>
   * ```
   */
  i18nInstance?: Streami18n;
  /**
   * You can pass the theme object to customize the styles of Chat components. You can check the default theme in [theme.ts](https://github.com/GetStream/stream-chat-react-native/blob/master/src/contexts/themeContext/utils/theme.ts)
   *
   * Please check section about [themes in cookbook](https://github.com/GetStream/stream-chat-react-native/wiki/Cookbook-v3.0#theme) for details.
   *
   * ```
   * import type { DeepPartial, Theme } from 'stream-chat-react-native';
   *
   * const theme: DeepPartial<Theme> = {
   *   messageSimple: {
   *     file: {
   *       container: {
   *         backgroundColor: 'red',
   *       },
   *       icon: {
   *         height: 16,
   *         width: 16,
   *       },
   *     },
   *   },
   * };
   *
   * <Chat style={theme}>
   * </Chat>
   * ```
   *
   * @overrideType object
   */
  style?: DeepPartial<Theme>;
};

const ChatWithContext = <
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType,
>(
  props: PropsWithChildren<ChatProps<At, Ch, Co, Ev, Me, Re, Us>>,
) => {
  const { children, client, closeConnectionOnBackground = true, i18nInstance, style } = props;

  const [channel, setChannel] = useState<Channel<At, Ch, Co, Ev, Me, Re, Us>>();
  const [translators, setTranslators] = useState<TranslationContextValue>({
    t: (key: string) => key,
    tDateTimeParser: (input?: string | number | Date) => Dayjs(input),
  });

  /**
   * Setup translators
   */
  const loadingTranslators = useStreami18n({ i18nInstance, setTranslators });

  /**
   * Setup connection event listeners
   */
  const { connectionRecovering, isOnline } = useIsOnline<At, Ch, Co, Ev, Me, Re, Us>(
    client,
    closeConnectionOnBackground,
  );

  /**
   * Setup muted user listener
   */
  const mutedUsers = useMutedUsers<At, Ch, Co, Ev, Me, Re, Us>(client);

  useEffect(() => {
    if (client.setUserAgent) {
      client.setUserAgent(`${SDK}-${Platform.OS}-${version}`);
      // This is to disable recovery related logic in js client, since we handle it in this SDK
      client.recoverStateOnReconnect = false;
    }
  }, []);

  const setActiveChannel = (newChannel?: Channel<At, Ch, Co, Ev, Me, Re, Us>) =>
    setChannel(newChannel);

  const chatContext = useCreateChatContext({
    channel,
    client,
    connectionRecovering,
    isOnline,
    mutedUsers,
    setActiveChannel,
  });

  if (loadingTranslators) return null;

  return (
    <ChatProvider<At, Ch, Co, Ev, Me, Re, Us> value={chatContext}>
      <TranslationProvider value={translators}>
        <ThemeProvider style={style}>{children}</ThemeProvider>
      </TranslationProvider>
    </ChatProvider>
  );
};

/**
 * Chat - Wrapper component for Chat. The needs to be placed around any other chat components.
 * This Chat component provides the ChatContext to all other components.
 *
 * The ChatContext provides the following props:
 *
 * - channel - currently active channel
 * - client - client connection
 * - connectionRecovering - whether or not websocket is reconnecting
 * - isOnline - whether or not set user is active
 * - setActiveChannel - function to set the currently active channel
 *
 * The Chat Component takes the following generics in order:
 * - At (AttachmentType) - custom Attachment object extension
 * - Ct (ChannelType) - custom Channel object extension
 * - Co (CommandType) - custom Command string union extension
 * - Ev (EventType) - custom Event object extension
 * - Me (MessageType) - custom Message object extension
 * - Re (ReactionType) - custom Reaction object extension
 * - Us (UserType) - custom User object extension
 */
export const Chat = <
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType,
>(
  props: PropsWithChildren<ChatProps<At, Ch, Co, Ev, Me, Re, Us>>,
) => {
  const { style } = useOverlayContext();

  return <ChatWithContext {...{ style }} {...props} />;
};
