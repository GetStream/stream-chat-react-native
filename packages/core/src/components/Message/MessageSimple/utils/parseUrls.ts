const protocol = '[a-z][\\w\\.+-]+:\\/\\/';
const hostName = '[w]{3,}\\.';
const domainName = '[\\w_-]+((?:\\.[\\w_-]+)+)';
const path = '[\\w\\p{Letter}()$/\\._-]+';
const queryString = '\\?[\\w\\p{Letter}=&+~_-]+';
const fragment = '#[\\w\\p{Letter}#?=&+~_-]+';
const port = ':[\\d]{2,5}';
const emailAllowedCharacters = '[\\w+\\.~$_-]+';

/**
 * Don't match the following pattern if the previous string
 * matches `[anything](` as we don't want to interfere
 * with links that already have been formatted as markdown
 * */
const notWithinLink = `(?<!\\[.+\\]\\(.+)`;

/**
 * Match any email address, with and without `mailto:`
 * */
const emailPattern = `(?:mailto:)?(?<emailDisplay>(?:${emailAllowedCharacters})@(?:${domainName}))`;
/**
 * Match links decided by having a protocol, allows for example
 * https://localhost to be recognized as a link
 * */
const protocolPrefixedLinkPattern = `(?<pplProtocol>${protocol})(?<pplDisplay>${notWithinLink}(?:${path})?(?:${queryString})?(?:${fragment})?)`;
/**
 * Match as much as possible of a fqdn, at least with a domain name formed as
 * something.tld
 * */
const fqdnPattern = `(?<fqdnProtocol>${protocol})?(?:${hostName})?(?<fqdnDisplay>${notWithinLink}(?:${domainName})(?:${port})?(?:${path})?(?:${queryString})?(?:${fragment})?)`;

const linkRegexp = new RegExp(
  `${emailPattern}|${fqdnPattern}|${protocolPrefixedLinkPattern}`,
  'giu',
);

interface Link {
  encoded: string;
  protocol: string;
  raw: string;
}

export const parseUrlsFromText = (input: string): Link[] => {
  let matches;

  const results: Array<{
    encoded: string;
    protocol: string;
    raw: string;
  }> = [];

  while ((matches = linkRegexp.exec(input)) !== null) {
    results.push(extractFromMatchGroups(matches[0], matches.groups ?? {}));
  }

  return results;
};

/**
 * We can't use the same names for multiple groups in a regexp,
 * this helper accepts all group names and builds a `Link`
 * object from it.
 * */
const extractFromMatchGroups = (
  rawLink: string,
  {
    emailDisplay = null,
    fqdnDisplay = null,
    fqdnProtocol = null,
    pplDisplay = null,
    pplProtocol = null,
  },
): Link => {
  if (emailDisplay !== null) {
    return {
      encoded: encodeURI(emailDisplay),
      protocol: 'mailto:',
      raw: rawLink,
    };
  }

  return {
    encoded: encodeURI(fqdnDisplay ?? pplDisplay ?? ''),
    protocol: fqdnProtocol ?? pplProtocol ?? '',
    raw: rawLink,
  };
};
