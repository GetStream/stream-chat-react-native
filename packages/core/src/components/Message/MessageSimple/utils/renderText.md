[src/components/Message/MessageSimple/utils/renderText.ts](https://github.com/GetStream/stream-chat-react-native/blob/master/src/components/Message/MessageSimple/utils/renderText.ts)

Function that handles markdown and parsing for the text of message.

Internally it uses [react-native-simple-markdown](https://github.com/GetStream/react-native-simple-markdown#styles-1) to handle markdown.

It accepts following params:

- markdownRules { object } - <https://github.com/GetStream/react-native-simple-markdown#rules>
- markdownStyles { object } - <https://github.com/GetStream/react-native-simple-markdown#styles-1>
- message { object } Message object
