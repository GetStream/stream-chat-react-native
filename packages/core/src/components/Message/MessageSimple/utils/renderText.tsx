import React from 'react';
import { GestureResponderEvent, Linking, Text } from 'react-native';
import anchorme from 'anchorme';
import truncate from 'lodash/truncate';
// @ts-expect-error
import Markdown from 'react-native-markdown-package';
import {
  DefaultRules,
  defaultRules,
  MatchFunction,
  ParseFunction,
  parseInline,
  ReactNodeOutput,
} from 'simple-markdown';

import type { MessageType } from '../../../MessageList/hooks/useMessageList';

import type {
  DefaultAttachmentType,
  DefaultChannelType,
  DefaultCommandType,
  DefaultEventType,
  DefaultMessageType,
  DefaultReactionType,
  DefaultUserType,
  UnknownType,
} from '../../../../types/types';
import type { MessageContextValue } from '../../../../contexts/messageContext/MessageContext';
import type { Colors, MarkdownStyle } from '../../../../contexts/themeContext/utils/theme';

const defaultMarkdownStyles: MarkdownStyle = {
  inlineCode: {
    fontSize: 13,
    padding: 3,
    paddingHorizontal: 5,
  },
  list: {
    marginBottom: 8,
    marginTop: 8,
  },
  listItemText: {
    flex: 0,
  },
  mentions: {
    fontWeight: '700',
  },
  // unfortunately marginVertical doesn't override the defaults for these within the 3rd party lib
  paragraph: {
    marginBottom: 8,
    marginTop: 8,
  },
  paragraphCenter: {
    marginBottom: 8,
    marginTop: 8,
  },
  paragraphWithImage: {
    marginBottom: 8,
    marginTop: 8,
  },
};

const parse: ParseFunction = (capture, parse, state) => ({
  content: parseInline(parse, capture[0], state),
});

export type MarkdownRules = Partial<DefaultRules>;

export type RenderTextParams<
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType,
> = Partial<
  Pick<MessageContextValue<At, Ch, Co, Ev, Me, Re, Us>, 'onLongPress' | 'onPress' | 'preventPress'>
> & {
  colors: typeof Colors;
  message: MessageType<At, Ch, Co, Ev, Me, Re, Us>;
  markdownRules?: MarkdownRules;
  markdownStyles?: MarkdownStyle;
  messageOverlay?: boolean;
  onLink?: (url: string) => Promise<void>;
  onlyEmojis?: boolean;
};

export const renderText = <
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType,
>(
  params: RenderTextParams<At, Ch, Co, Ev, Me, Re, Us>,
) => {
  const {
    colors,
    markdownRules,
    markdownStyles,
    message,
    messageOverlay,
    onLink: onLinkParams,
    onLongPress: onLongPressParam,
    onlyEmojis,
    onPress: onPressParam,
    preventPress,
  } = params;

  // take the @ mentions and turn them into markdown?
  // translate links
  const { mentioned_users, text } = message;

  if (!text) return null;

  let newText = text.trim();
  const urls = anchorme(newText, {
    list: true,
  });

  for (const urlInfo of urls) {
    const displayLink = truncate(urlInfo.encoded.replace(/^(www\.)/, ''), {
      length: 200,
      omission: '...',
    });
    const markdown = `[${displayLink}](${urlInfo.protocol}${urlInfo.encoded})`;
    newText = newText.replace(urlInfo.raw, markdown);
  }

  newText = newText.replace(/[<&"'>]/g, '\\$&');
  const styles: MarkdownStyle = {
    ...defaultMarkdownStyles,
    ...markdownStyles,
    autolink: {
      ...defaultMarkdownStyles.autolink,
      color: colors.accent_blue,
      ...markdownStyles?.autolink,
    },
    inlineCode: {
      ...defaultMarkdownStyles.inlineCode,
      backgroundColor: colors.white_smoke,
      borderColor: colors.grey_gainsboro,
      color: colors.accent_red,
      ...markdownStyles?.inlineCode,
    },
    mentions: {
      ...defaultMarkdownStyles.mentions,
      color: colors.accent_blue,
      ...markdownStyles?.mentions,
    },
    text: {
      ...defaultMarkdownStyles.text,
      color: colors.black,
      ...markdownStyles?.text,
    },
  };

  const onLink = (url: string) =>
    onLinkParams
      ? onLinkParams(url)
      : Linking.canOpenURL(url).then((canOpenUrl) => canOpenUrl && Linking.openURL(url));

  const react: ReactNodeOutput = (node, output, { ...state }) => {
    const onPress = (event: GestureResponderEvent) => {
      if (!preventPress && onPressParam) {
        onPressParam({
          defaultHandler: () => onLink(node.target),
          emitter: 'textLink',
          event,
        });
      }
    };

    const onLongPress = (event: GestureResponderEvent) => {
      if (!preventPress && onLongPressParam) {
        onLongPressParam({
          emitter: 'textLink',
          event,
        });
      }
    };

    state.withinLink = true;
    const link = React.createElement(
      Text,
      {
        key: state.key,
        onLongPress,
        onPress,
        style: styles.autolink,
        suppressHighlighting: true,
      },
      output(node.content, state),
    );
    state.withinLink = false;
    return link;
  };

  const mentionedUsers = Array.isArray(mentioned_users)
    ? mentioned_users.reduce((acc, cur) => {
        const userName = cur.name || cur.id || '';
        if (userName) {
          acc += `${acc.length ? '|' : ''}@${userName}`;
        }
        return acc;
      }, '')
    : '';

  const regEx = new RegExp(`^\\B(${mentionedUsers})`, 'g');
  const match: MatchFunction = (source) => regEx.exec(source);

  const mentionsReact: ReactNodeOutput = (node, output, { ...state }) => {
    const onPress = (event: GestureResponderEvent) => {
      if (!preventPress && onPressParam) {
        onPressParam({
          emitter: 'textMention',
          event,
        });
      }
    };

    const onLongPress = (event: GestureResponderEvent) => {
      if (!preventPress && onLongPressParam) {
        onLongPressParam({
          emitter: 'textMention',
          event,
        });
      }
    };

    return React.createElement(
      Text,
      {
        key: state.key,
        onLongPress,
        onPress,
        style: styles.mentions,
      },
      Array.isArray(node.content) ? node.content[0]?.content || '' : output(node.content, state),
    );
  };

  const customRules = {
    link: { react },
    // we have no react rendering support for reflinks
    reflink: { match: () => null },
    ...(mentionedUsers
      ? {
          mentions: {
            match,
            order: defaultRules.text.order - 0.5,
            parse,
            react: mentionsReact,
          },
        }
      : {}),
  };

  return (
    <Markdown
      key={`${JSON.stringify(mentioned_users)}-${onlyEmojis}-${
        messageOverlay ? JSON.stringify(markdownStyles) : undefined
      }`}
      onLink={onLink}
      rules={{
        ...customRules,
        ...markdownRules,
      }}
      styles={styles}
    >
      {newText}
    </Markdown>
  );
};
