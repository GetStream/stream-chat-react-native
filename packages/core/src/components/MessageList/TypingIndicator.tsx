import React from 'react';
import { StyleSheet, Text, View } from 'react-native';

import { useTypingString } from './hooks/useTypingString';

import { useTheme } from '../../contexts/themeContext/ThemeContext';

import type {
  DefaultAttachmentType,
  DefaultChannelType,
  DefaultCommandType,
  DefaultEventType,
  DefaultMessageType,
  DefaultReactionType,
  DefaultUserType,
  UnknownType,
} from '../../types/types';
import { LoadingDots } from '../Indicators/LoadingDots';

const styles = StyleSheet.create({
  container: {
    alignItems: 'center',
    flexDirection: 'row',
    height: 24,
    justifyContent: 'flex-start',
  },
  loadingDots: {
    marginLeft: 8,
  },
  typingText: {
    marginLeft: 8,
  },
});

export const TypingIndicator = <
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends DefaultUserType = DefaultUserType,
>() => {
  const {
    theme: {
      colors: { grey, white_snow },
      typingIndicator: { container, text },
    },
  } = useTheme();
  const typingString = useTypingString<At, Ch, Co, Ev, Me, Re, Us>();

  return (
    <View
      style={[styles.container, { backgroundColor: `${white_snow}E6` }, container]}
      testID='typing-indicator'
    >
      <LoadingDots style={styles.loadingDots} />
      <Text style={[styles.typingText, { color: grey }, text]}>{typingString}</Text>
    </View>
  );
};

TypingIndicator.displayName = 'TypingIndicator{typingIndicator}';
