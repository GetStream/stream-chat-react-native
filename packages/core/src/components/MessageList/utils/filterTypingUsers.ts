import type { ChatContextValue } from '../../../contexts/chatContext/ChatContext';
import type { ThreadContextValue } from '../../../contexts/threadContext/ThreadContext';
import type { TypingContextValue } from '../../../contexts/typingContext/TypingContext';
import type {
  DefaultAttachmentType,
  DefaultChannelType,
  DefaultCommandType,
  DefaultEventType,
  DefaultMessageType,
  DefaultReactionType,
  DefaultUserType,
  UnknownType,
} from '../../../types/types';

type FilterTypingUsersParams<
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType,
> = Pick<TypingContextValue<At, Ch, Co, Ev, Me, Re, Us>, 'typing'> &
  Pick<ChatContextValue<At, Ch, Co, Ev, Me, Re, Us>, 'client'> &
  Pick<ThreadContextValue<At, Ch, Co, Ev, Me, Re, Us>, 'thread'>;

export const filterTypingUsers = <
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType,
>({
  client,
  thread,
  typing,
}: FilterTypingUsersParams<At, Ch, Co, Ev, Me, Re, Us>) => {
  const nonSelfUsers: string[] = [];

  if (!client || !client.user || !typing) return nonSelfUsers;

  const typingKeys = Object.keys(typing);

  typingKeys.forEach((typingKey) => {
    if (!typing[typingKey]) return;

    /** removes own typing events */
    if (client.user?.id === typing[typingKey].user?.id) {
      return;
    }

    const isRegularEvent = !typing[typingKey].parent_id && !thread?.id;
    const isCurrentThreadEvent = typing[typingKey].parent_id === thread?.id;

    /** filters different threads events */
    if (!isRegularEvent && !isCurrentThreadEvent) {
      return;
    }

    const user = typing[typingKey].user?.name || typing[typingKey].user?.id;
    if (user) {
      nonSelfUsers.push(user);
    }
  });

  return nonSelfUsers;
};
