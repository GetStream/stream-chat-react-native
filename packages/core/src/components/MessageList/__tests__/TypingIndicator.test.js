import React from 'react';
import { cleanup, render, waitFor } from '@testing-library/react-native';

import { TypingIndicator } from '../TypingIndicator';

import { Chat } from '../../Chat/Chat';

import { TypingProvider } from '../../../contexts/typingContext/TypingContext';
import { TranslationProvider } from '../../../contexts/translationContext/TranslationContext';
import { Streami18n } from '../../../utils/Streami18n';

import { generateStaticUser, generateUser } from '../../../mock-builders/generator/user';
import { getTestClientWithUser } from '../../../mock-builders/mock';

afterEach(cleanup);

describe('TypingIndicator', () => {
  let chatClient;

  it('should render typing indicator for two users', async () => {
    const t = jest.fn((key) => key);
    const user0 = generateUser();
    const user1 = generateUser();
    const user2 = generateUser();

    chatClient = await getTestClientWithUser({ id: user0.id });
    await chatClient.setUser(user0, 'testToken');
    const typing = { user1: { user: user1 }, user2: { user: user2 } };

    const { getByTestId } = render(
      <Chat client={chatClient}>
        <TranslationProvider value={{ t }}>
          <TypingProvider value={{ typing }}>
            <TypingIndicator />
          </TypingProvider>
        </TranslationProvider>
      </Chat>,
    );
    expect(t).toHaveBeenCalledWith('{{ firstUser }} and {{ nonSelfUserLength }} more are typing', {
      firstUser: user1.name,
      nonSelfUserLength: 1,
    });
    await waitFor(() => {
      expect(getByTestId('typing-indicator')).toBeTruthy();
    });
  });

  it('should render typing indicator for one user', async () => {
    const t = jest.fn((key) => key);
    const user0 = generateUser();
    const user1 = generateUser();

    chatClient = await getTestClientWithUser({ id: user0.id });
    await chatClient.setUser(user0, 'testToken');
    const typing = { user1: { user: user1 } };

    const { getByTestId } = render(
      <Chat client={chatClient}>
        <TranslationProvider value={{ t }}>
          <TypingProvider value={{ typing }}>
            <TypingIndicator />
          </TypingProvider>
        </TranslationProvider>
      </Chat>,
    );
    expect(t).toHaveBeenCalledWith('{{ user }} is typing', {
      user: user1.name,
    });
    await waitFor(() => {
      expect(getByTestId('typing-indicator')).toBeTruthy();
    });
  });

  it('should match typing indicator snapshot', async () => {
    const i18nInstance = new Streami18n();
    const { t } = await i18nInstance.getTranslators();
    const user0 = generateStaticUser(0);
    const user1 = generateStaticUser(1);
    const user2 = generateStaticUser(3);

    chatClient = await getTestClientWithUser({ id: user0.id });
    await chatClient.setUser(user0, 'testToken');
    const typing = { user1: { user: user1 }, user2: { user: user2 } };

    const { toJSON } = render(
      <Chat client={chatClient}>
        <TranslationProvider value={{ t }}>
          <TypingProvider value={{ typing }}>
            <TypingIndicator />
          </TypingProvider>
        </TranslationProvider>
      </Chat>,
    );
    await waitFor(() => {
      expect(toJSON()).toMatchSnapshot();
    });
  });
});
