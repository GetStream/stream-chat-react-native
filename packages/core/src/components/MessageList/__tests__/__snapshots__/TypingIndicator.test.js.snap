// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TypingIndicator should match typing indicator snapshot 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "height": 24,
        "justifyContent": "flex-start",
      },
      Object {
        "backgroundColor": "#FCFCFCE6",
      },
      Object {},
    ]
  }
  testID="typing-indicator"
>
  <View
    style={
      Array [
        Object {
          "flexDirection": "row",
        },
        Object {},
        Object {
          "marginLeft": 8,
        },
      ]
    }
  >
    <View
      animatedStyle={
        Object {
          "value": Object {
            "opacity": -0.3333333333333333,
          },
        }
      }
      collapsable={false}
      style={
        Object {
          "backgroundColor": "#000000",
          "borderRadius": 2,
          "height": 4,
          "marginRight": 2,
          "opacity": -0.3333333333333333,
          "width": 4,
        }
      }
    />
    <View
      animatedStyle={
        Object {
          "value": Object {
            "opacity": 0.3333333333333333,
          },
        }
      }
      collapsable={false}
      style={
        Object {
          "backgroundColor": "#000000",
          "borderRadius": 2,
          "height": 4,
          "marginHorizontal": 2,
          "opacity": 0.3333333333333333,
          "width": 4,
        }
      }
    />
    <View
      animatedStyle={
        Object {
          "value": Object {
            "opacity": 1,
          },
        }
      }
      collapsable={false}
      style={
        Object {
          "backgroundColor": "#000000",
          "borderRadius": 2,
          "height": 4,
          "marginLeft": 2,
          "opacity": 1,
          "width": 4,
        }
      }
    />
  </View>
  <Text
    style={
      Array [
        Object {
          "marginLeft": 8,
        },
        Object {
          "color": "#7A7A7A",
        },
        Object {
          "fontSize": 14,
        },
      ]
    }
  >
    Grant and 1 more are typing
  </Text>
</View>
`;
